servicetalk-transport-netty-internal/src/main/java/io/servicetalk/transport/netty/internal/TransportObserverUtils.java

```diff
@@ -68,25 +83,84 @@ static ConnectionObserver connectionObserver(final Channel channel) {
      * @param channel a {@link Channel} to assign a {@link Throwable} to
      * @param error a {@link Throwable}
      */
-    static void assignConnectionError(final Channel channel, final Throwable error) {
+    public static void assignConnectionError(final Channel channel, final Throwable error) {
         if (connectionObserver(channel) != null) {
             channel.attr(CONNECTION_ERROR).setIfAbsent(error);
         }
     }
 
+    /**
+     * Returns an {@link Throwable error} assigned with the specified {@link Channel}.
+     *
+     * @param channel to look for a {@link Throwable}
+     * @return an {@link Throwable error} assigned with the specified {@link Channel}.
+     */
     @Nullable
-    private static Throwable connectionError(final Channel channel) {
+    public static Throwable connectionError(final Channel channel) {
         return channel.attr(CONNECTION_ERROR).getAndSet(null);
     }
 
     static void reportSecurityHandshakeStarting(final Channel channel) {
         final ConnectionObserver observer = connectionObserver(channel);
         assert observer != null;
-        channel.attr(SECURITY_HANDSHAKE_OBSERVER).set(requireNonNull(observer.onSecurityHandshake()));
+        channel.attr(SECURITY_HANDSHAKE_OBSERVER)
+                .set(safeReport(observer::onSecurityHandshake, observer, "security handshake"));
     }
 
     @Nullable
     static SecurityHandshakeObserver securityHandshakeObserver(final Channel channel) {
         return channel.attr(SECURITY_HANDSHAKE_OBSERVER).get();
     }
+
+    /**
+     * Safely executes the passed {@link Runnable} and logs unexpected exception if any.
+     *
+     * @param runnable {@link Runnable} to run
+     * @param observer {@link Object} that is invoked
+     * @param event event that is invoked
+     */
+    public static void safeReport(final Runnable runnable, final Object observer, final String event) {
```

Comment:
An alternate to having call sites worry about unexpected error handling is to wrap the `Observer` with a catch-all implementation. The positives are:
- Less pollution on every call-site and danger of missing the error handling at some call-sites.
- Central place where certain observer implementations can avoid allocation if they are known to be safe.

The cons are arguably higher allocation which is debatable when compared to this approach which allocates a new `Runnable` (capturing lambdas mostly) on each call.

Did you consider this alternative?

Reply:
Yes, I considered this alternative and do not worry much about the allocation overhead of catch-all implementation.
The case that I couldn't solve with that approach is when the original observer unexpectedly returns `null`, instead of another observer layer. If I do
```java
final class CatchAllTransportObserver implements TransportObserver {
    //...
    @Override
    public ConnectionObserver onNewConnection() {
        final ConnectionObserver connectionObserver = safeReport(delegate::onNewConnection, delegate, "new connection");
        if (connectionObserver == null) {
            return null;
        }
        return new CatchAllConnectionObserver(connectionObserver);
    }
```
Then intellij idea and spotbugs both complain:
```
io.servicetalk.transport.netty.internal.CatchAllTransportObserver.onNewConnection() may return null, but is declared @Nonnull
```

Also, the caller of an observer API should always do something like:
```java
        final ConnectionObserver observer = transportObserver.onNewConnection();
        if (observer == null) {
            return;
        }
```
where intellij idea and spotbugs again complain:
```
Redundant nullcheck of observer, which is known to be non-null in io.servicetalk.transport.netty.internal.TransportObserverInitializer.init(Channel)
```

I decided that it will be annoying to suppress all these warnings twice: in spotbugs exclusion XML and an intellij idea via comments/annotations. And don't think that we need to annotate all methods as `@Nullable`.

LMK what you think and I can address it in a follow-up.

Reply:
```
if (connectionObserver == null) {
            return null;
}
```

return a `NoopConnectionObserver` instead of `null`?